src/logics/Theory.cc

```diff
@@ -52,26 +52,27 @@ Theory::SubstitutionResult Theory::computeSubstitutions(const PTRef fla)
         // Add the newly obtained units to the list of all substitutions
         // Clear the previous units
         new_units.getKeysAndVals(new_units_vec);
-        for (int i = 0; i < new_units_vec.size(); i++) {
-            Map<PTRef,lbool,PTRefHash>::Pair unit = new_units_vec[i];
+        for (auto unit : new_units_vec) {
             current_units_vec.push(PtAsgn(unit.key, unit.data));
         }
         Map<PTRef,PtAsgn,PTRefHash> newsubsts;
         lbool res = getLogic().retrieveSubstitutions(current_units_vec, newsubsts);
-        getLogic().substitutionsTransitiveClosure(newsubsts);
+        getLogic().substitutionsTransitiveClosure<PtAsgn>(newsubsts);
         if (res != l_Undef)
             root = (res == l_True ? getLogic().getTerm_true() : getLogic().getTerm_false());
-        PTRef new_root = Substitutor(getLogic(), newsubsts).rewrite(root);
-        bool cont = new_root != root;
-        // remember the substitutions
-        vec<Map<PTRef,PtAsgn,PTRefHash>::Pair> newsubsts_vec;
-        newsubsts.getKeysAndVals(newsubsts_vec);
-        for (int i = 0; i < newsubsts_vec.size(); ++i) {
-            PTRef key = newsubsts_vec[i].key;
-            if (!allsubsts.has(key) && newsubsts_vec[i].data.sgn == l_True) {
-                allsubsts.insert(key, newsubsts_vec[i].data);
+        auto newRootAndSubsts(Substitutor<PtAsgn>(getLogic(), newsubsts).rewrite(root));
+        PTRef new_root = newRootAndSubsts.first;
+        auto usedSubstitutions(std::move(newRootAndSubsts.second));
```

Comment:
Here's essentially why I felt that I need to change the interface of the Substitutor.  The substitutions that are computed are used by the models (and indeed also by this very substitution loop).  It might be good to write some templates to avoid having to deal every time with the returned substitution map in places where it's not needed.